first draft with lfp conversion stubs, folder structure and class def…

h-mayorquin opened the pull request:
Work in progress.

First draft with both the converter script and the converter class.

To-do:
- [x] Add lfp data in the converter script.
- [x] Figure out for which files raw data exists and add them. 
- [x] Add recording information contained in XML files.
- [x] Check if `clu` and `res` match the `cell.info` spike information.
- [x] Add spike interface to the converter script.
- [x] Describe the matlab files available and add them to the documentation https://github.com/catalystneuro/buzsaki-lab-to-nwb/pull/38.
- [x] Check if the cell explorer interface is compatible with the files available.
- [x] Make a list of behaviors to add and the files where they are contained/described/recorded.
- [x] Add the behavioral interface for the list above.
- [x] Add metadata.
- [ ] Run conversion stubs and review nwb files to check for right format, compression schemes, types and sizes.
- [ ] Run full conversion.
- [ ] Review and finish documentation so it is align with converter https://github.com/catalystneuro/buzsaki-lab-to-nwb/pull/38.

CodyCBakerPhD commented:
Thanks so much for opening the PR on this! And for making a to-do list - keep up the good work!

bendichter commented:
@h-mayorquin yes, this looks great. FYI, you are able to open PRs in "draft mode" which formally indicates that they are work-in-progress. You can then open it for review once it is ready. This is my preferred approach because it opens your work to the rest of the team while making it clear that it is not finished yet. I have changed this one for you.

bendichter reviewed (commented):
looks good so far!

h-mayorquin commented:
> while

Great, will do this from now on.

h-mayorquin commented:
I have some questions about adding the interface for the spike data.

For this data repository what we have available are `.clu` and `.res` files. They come in more than one per session and follow the convention `{session_name}.{clu|res}.{x}` where x is an integer. Searching through the conversion tools repository it seems that `NeuroscopeMultiSortingExtractor` is the right tool for this case. 

However, the problem is that `NeuroscopeMultiSortingExtractor` expects a single `.xml` file per session and some of them have two.  As an example, subject `DT5` session's `20161021_1840um_2412um_merge` contains two xml files:
1) `20161021_1840um_2412um_merge.xml`
2) `rips.rls.xml` 

My guess is that the first one is correct but the spiking interface throws the following assertion:
`AssertionError: More than one .xml file found in the folder!`

How do we deal with this cases? A simple solution is just to remove the non useful xml if it is present in some session as we seem to not be using them.

Another solution is to modify the code in the `NeuroscopeMultiSortingExtractor` class to add an optional `xml_file_path` as an argument. This is available in  `NeuroscopeRecordingExtractor` and maybe should be here as well?

Have you encounter this issue previously, @CodyCBakerPhD? Is there an agreed upon solution or am I missing something?

Also, for this data, we do have wave forms available (`.spk` files) should we use the parameters `load_wave_forms` to get them or do we usually skip them?

bendichter commented:
Let's add an argument for xlm_file_path

On Thu, Nov 18, 2021 at 3:52 AM Heberto Mayorquin ***@***.***>
wrote:

> I have some questions about adding the interface for the spike data.
>
> For this data repository what we have available are .clu and .res files.
> They come in more than one per session and follow the convention
> {session_name}.{clu|res}.{x} where x is an integer. Searching through the
> conversion tools repository it seems that NeuroscopeMultiSortingExtractor
> is the right tool for this case.
>
> However, the problem is that NeuroscopeMultiSortingExtractor expects a
> single .xml file per session and some of them have two. As an example,
> subject DT5 session's 20161021_1840um_2412um_merge contains two xml files:
>
>    1. 20161021_1840um_2412um_merge.xml
>    2. rips.rls.xml
>
> My guess is that the first one is correct but the spiking interface throws
> the following assertion:
> AssertionError: More than one .xml file found in the folder!
>
> How do we deal with this cases? A simple solution is just to remove the
> non useful xml if it is present in some session as we seem to not be using
> them.
>
> Another solution is to modify the code in the
> NeuroscopeMultiSortingExtractor class to add an optional xml_file_path as
> an argument. This is available in NeuroscopeRecordingExtractor and maybe
> should be here as well?
>
> Have you encounter this issue previously, @CodyCBakerPhD
> <https://github.com/CodyCBakerPhD>? Is there an agreed upon solution or
> am I missing something?
>
> Also, for this data, we do have wave forms available (.spk files) should
> we use the parameters load_wave_forms to get them or do we usually skip
> them?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/catalystneuro/buzsaki-lab-to-nwb/pull/39#issuecomment-972706720>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGOEETMR5LS2BIVNOAKXVTUMTEG7ANCNFSM5IBI2TPA>
> .
>


-- 
Ben Dichter, PhD
Data Science Consultant
personal website <http://bendichter.com/>

CodyCBakerPhD commented:
> Also, for this data, we do have wave forms available (`.spk` files) should we use the parameters `load_wave_forms` to get them or do we usually skip them?

Skip, until we enable conversion factors for waveform data. Otherwise, they'll take up 4 times as much space.

> Let's add an argument for xlm_file_path

Yep, I think @h-mayorquin and I actually noticed this in the last dataset. I'll update the extractor and interface today.

CodyCBakerPhD commented:
@h-mayorquin Do all sessions have both CellExplorer and Neuroscope? If so, you could also just check a couple sessions manually by temporarily removing that other .xml file and see if the two formats have the same spikes, just like we did last time. If so we can also just ignore the Neuroscope in that case and default to using CellExplorer.

CodyCBakerPhD commented:
@h-mayorquin Either way I guess, check out [this PR](https://github.com/catalystneuro/nwb-conversion-tools/pull/297) and the spikeextractors PR it relies on to see if that fixes the issue for you

h-mayorquin commented:
> @h-mayorquin Do all sessions have both CellExplorer and Neuroscope? If so, you could also just check a couple sessions manually by temporarily removing that other .xml file and see if the two formats have the same spikes, just like we did last time. If so we can also just ignore the Neuroscope in that case and default to using CellExplorer.

No, I just added this to the documentation pull request. Around ~ 33 % of the sessions have neither Nerusocope nor CellExplorer files (not `clu`, `res` or `spikes.cellinfo.mat` files).

I just realized that git does not quote in full? I am not sure if you get a reference when I quote-reply you @CodyCBakerPhD .

CodyCBakerPhD commented:
That is very strange. Do any of those 33% of sessions have any other discernable type of spike sorting? or is it just raw data and behavior? Can you send me a session ID for one of those and I can also take a look myself.

> I just realized that git does not quote in full? I am not sure if you get a reference when I quote-reply you @CodyCBakerPhD .

I get a notification no matter what because I 'watch' all activity on this repo. So either way; but Git quoting is just a way of nesting any text occurring after a '>', it doesn't actually link to the previous comments

h-mayorquin commented:
> That is very strange. Do any of those 33% of sessions have any other discernable type of spike sorting? or is it just raw data and behavior? Can you send me a session ID for one of those and I can also take a look myself.

Yes, the seem to be only raw data. I am not sure if they have behavior. Some of the ones I just opened randomly only have session.info, the raw data and nothing else.
Here are some:
  
    /shared/catalystneuro/Buzsaki/TingleyD/DT8/20170124_0um_0um_170124_151501
    /shared/catalystneuro/Buzsaki/TingleyD/DT8/20170228_324um_2088um_merge
    /shared/catalystneuro/Buzsaki/TingleyD/DT8/20170228_324um_2088um_merge

h-mayorquin commented:
> @h-mayorquin Either way I guess, check out [this PR](https://github.com/catalystneuro/nwb-conversion-tools/pull/297) and the spikeextractors PR it relies on to see if that fixes the issue for you

I tested this and it seems to work in the sense that it does not produce error (note, what I did was to git clone the nwb-conversion-tools repository, switch the branch and install that version with pip, maybe there is a a better way?):

```python
  from nwb_conversion_tools import NeuroscopeSortingInterface
  
  session_path = Path('/shared/catalystneuro/Buzsaki/TingleyD/DT8/20170201_72um_432um_merge')
  #session_path = Path('/shared/catalystneuro/Buzsaki/TingleyD/DT2/DT2_rPPC_rCCG_144um_72um_20160207_160207_170623')
  clu_matches_in_session = len(list(session_path.glob("*.clu*")))
  res_matches_in_session = len(list(session_path.glob("*.res*")))
  xml_file_path = session_path.joinpath(session_path.stem +'.xml')
  print(f" xml is a file {xml_file_path.is_file()}")
  print(f" clu matches in session {clu_matches_in_session}")
  print(f" res matches in session {res_matches_in_session}")

  multisorting_neuroscope = NeuroscopeSortingInterface(folder_path=session_path, keep_mua_units=False, xml_file_path=str(xml_file_path))
  multisorting_neuroscope.get_metadata()  # ElectrodeGroup contained here `
```
This runs without error, and the output is along the following lines

```json
  {'ElectrodeGroup': [{'name': 'Group1',
     'description': 'Group1 electrodes.',
     'location': '',
     'device': 'Device_ecephys'},
    {'name': 'Group2',
     'description': 'Group2 electrodes.',
     'location': '',
     'device': 'Device_ecephys'},
    {'name': 'Group3',
     'description': 'Group3 electrodes.',
```


I mention this because when I try to **add this to the converter script and run it**, there seems to be a problem with the metadata. I get the following error:
`jsonschema.exceptions.ValidationError: Additional properties are not allowed ('Electrodes', 'ElectrodeGroup', 'UnitProperties' were unexpected)`

Any idea what could be going on? I am not sure how I track this one but I just encountered it, I would give it some thought. 
The converter script in the last commit contains the code to reproduce the error as commented lines if you want to take a look at it:

https://github.com/catalystneuro/buzsaki-lab-to-nwb/blob/f2660b0b9a3ca81e2ccd2ccec21a4ecab274afdc/buzsaki_lab_to_nwb/tingley_code_septal/convert_tingley_septal.py#L53-L60

h-mayorquin commented:
Following Ben's recommendation on the group discussion yesterday, here is the complete error trace for the problem described in my previous comment:
https://gist.github.com/h-mayorquin/458a950c7cc32d7d9047cb4096992460

There seems to a level mismatch between the metadata extracted from `NeuroscopeSortingInterface` and the one expected from the converter. I am still unable to figure this out.

h-mayorquin commented:
OK, so I managed to understand the scheme and I think is a mismatch. Check the latest commit to [the xml PR](https://github.com/catalystneuro/nwb-conversion-tools/pull/297) that @CodyCBakerPhD opened.

h-mayorquin commented:
I added new comments as the data interface for the spiking data is still producing errors ( to https://github.com/catalystneuro/nwb-conversion-tools/pull/297).

Moreover, after using the latest version of `nwb_conversion_tools` and the [branch](https://github.com/SpikeInterface/spikeextractors/pull/665) `spikeextractors` that @CodyCBakerPhD started for the other bug I found I am also getting errors for the raw data file:

```python
  /shared/catalystneuro/Buzsaki/TingleyD/DT2/DT2_rPPC_rCCG_3612um_1360um_20160302_160302_134841
  raw file available True
  lfp file available True
  Source data is valid!
  Metadata is valid!
  conversion_options is valid!
  /home/jovyan/development/nwb-conversion-tools/nwb_conversion_tools/utils/basenwbephyswriter.py:210: UserWarning: cannot create electrodes for this recording as ids already exist
    warnings.warn("cannot create electrodes for this recording as ids already exist")
  Traceback (most recent call last):
    File "buzsaki_lab_to_nwb/tingley_code_septal/convert_tingley_septal.py", line 76, in <module>
      overwrite=True,
    File "/home/jovyan/development/nwb-conversion-tools/nwb_conversion_tools/nwbconverter.py", line 180, in run_conversion
      data_interface.run_conversion(nwbfile, metadata, **conversion_options.get(interface_name, dict()))
    File "/home/jovyan/development/nwb-conversion-tools/nwb_conversion_tools/datainterfaces/ecephys/baserecordingextractorinterface.py", line 141, in run_conversion
      self.writer_class.add_to_nwb(nwbfile, metadata, **conversion_opts)
    File "/home/jovyan/development/nwb-conversion-tools/nwb_conversion_tools/utils/basesinwbephyswriter.py", line 70, in add_to_nwb
      self.add_recording()
    File "/home/jovyan/development/nwb-conversion-tools/nwb_conversion_tools/utils/si013nwbephyswriter.py", line 229, in add_recording
      super().add_recording(segment_index=segment_index)
    File "/home/jovyan/development/nwb-conversion-tools/nwb_conversion_tools/utils/basesinwbephyswriter.py", line 50, in add_recording
      self.add_electrical_series(segment_index=segment_index)
    File "/home/jovyan/development/nwb-conversion-tools/nwb_conversion_tools/utils/basenwbephyswriter.py", line 391, in add_electrical_series
      ), f"Raw ElectricalSeries '{eseries_kwargs['name']}' is already written in the NWBFile!"
  AssertionError: Raw ElectricalSeries 'ElectricalSeries' is already written in the NWBFile!

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/convert_tingley_septal.py:
```diff
@@ -0,0 +1,88 @@
+    )
+    converter = TingleySeptalNWBConverter(source_data)
+
+    metadata = None
```
```suggestion
```

CodyCBakerPhD commented:
@h-mayorquin That definitely looks like a `nwb-conversion-tools v0.10.1` error. Something with the default naming of multiple ElectricalSeries. I'll look into it but a quick workaround would be to manually control the names through the metadata like we did in the last project.

h-mayorquin commented:
There is another schema validation error with the `CellExplorer` metadata. Here is the trace for the session `"/shared/catalystneuro/Buzsaki/TingleyD/DT8/20170220_216um_1944um_170220_192456"`:

```python
  ValidationError: {'name': 'location', 'description': 'Brain region where each unit was detected.'} is not of type 'array'
  
  Failed validating 'type' in schema['properties']['Ecephys']['properties']['UnitProperties']:
      {'default': {'description': 'Brain region where each unit was '
                                  'detected.',
                   'name': 'location'},
       'items': {'$ref': '#/properties/Ecephys/properties/definitions/UnitProperties'},
       'minItems': 0,
       'renderForm': False,
       'type': 'array'}
  
  On instance['Ecephys']['UnitProperties']:
      {'description': 'Brain region where each unit was detected.',
       'name': 'location'}
```
To my understanding the error means that the schema expected an array (in this case a list) for `UnitProperties` but is getting a dic.
I did some introspection to see where this mismatch is originated. Two important pieces of evidence:
* When I use the `get_metadata()` method of the `CellExplorerSortingInterface` data interface by itself, it returns the correct format of `UnitProperties` as a list. 
* Moreover, the error is not triggered when I run the convert script using only the `CellExplorerSortingInterface` (i.e. without any other interface).

My guess, is that the [deep dictionary update ](https://github.com/catalystneuro/nwb-conversion-tools/blob/80f53cf86c86f3055b5bca328a841ec9e2fbe1e4/nwb_conversion_tools/nwbconverter.py#L91-L97 ) in `nwb conversion-tools` is the responsible for this.

Here is a gist with a minimal example that illustrates the problem:
https://gist.github.com/h-mayorquin/bc2556b6105b1b6715f5ec1474a1a443

CodyCBakerPhD commented:
@h-mayorquin The minimal example fails in two different ways - first because the order of the update is not applied to the more general metadata - changing it to `dict_deep_update(metadata_from_cellexplorer_sorting, metadata_from_neuroscope_lfp)` leads to the expected behavior and fixes the outer `NWBFile` key as well.

Note the first example can also be fixed by simply setting `dict_deep_update(metadata_from_neuroscope_lfp, metadata_from_cellexplorer_sorting, append_list=False)` which returns `UnitProperties` as the appropriate list.

I'll look for the best place to introduce this fix into the tools, but in the meantime can you raise the `deep_dict_update` issue (where `append_list=True` is not returning the expected behavior) on `nwb-conversion-tools` and ping Saksham about it? Looks like something that came out of some recent work he did on the function.

CodyCBakerPhD commented:
@h-mayorquin It also occurs to me, based on above, you *might* be able to locally bypass the update issue by defining all `CellExplorerSortingInterface` arguments in any dictionaries (`data_interface_classes`, `source_data`, etc.) before the Recording/LFP's do.

h-mayorquin commented:
> @h-mayorquin The minimal example fails in two different ways - first because the order of the update is not applied to the more general metadata - changing it to `dict_deep_update(metadata_from_cellexplorer_sorting, metadata_from_neuroscope_lfp)` leads to the expected behavior and fixes the outer `NWBFile` key as well.
> 
@CodyCBakerPhD 
Running the dic_deep_update with the arguments inverted seems to ignore most of the items in the group electrodes key (it only gets the last one).

I opened the issue: https://github.com/catalystneuro/nwb-conversion-tools/issues/314

h-mayorquin commented:
> @h-mayorquin It also occurs to me, based on above, you _might_ be able to locally bypass the update issue by defining all `CellExplorerSortingInterface` arguments in any dictionaries (`data_interface_classes`, `source_data`, etc.) before the Recording/LFP's do.

Yes, I tried this naively and it did not work. Writing the data interface dic in the NWB instantiation script does accomplish this but it generates a different error. 

Anyway, the solution seems brittle to me and it was out of curiosity. The `data_interface_objects` attribute of the  `NWBConverter` object is just a dic and we should not expect ordered based iteration from them, right?

CodyCBakerPhD commented:
> The data_interface_objects attribute of the NWBConverter object is just a dic and we should not expect ordered based iteration from them, right?

Yeah, the order should never matter, I was just curious if that would work though. Definitely an issue with the deep update

h-mayorquin commented:
With the latest bug fixes all the data transformations (raw, lfp and spikes) are working. Next is behavioral data.

h-mayorquin commented:
Behavioral data is now added. I will check with @CodyCBakerPhD  today to see if we are missing something, discuss best practices and code review.

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,131 @@
+            warnings.warn(f"Units not available")
+            units = None
+            conversion = np.nan
+        try:
+            description = behavior_mat["description"].replace("/", "")
+        except KeyError:
+            description = "generic_position_tracking"
```
```suggestion
            description = behavior_mat.get("description", "").replace("/", "")
```
`Try`/`except` catching is really only for situations where you can't do an `if` check or some other safe operation. It shouldn't be a general habit, even with non-bare `except`.

Can you change the rest of these similarly?

CodyCBakerPhD replied:
The second argument of `dictionary_object.get("key_name", "default")` is what is returned if the `key_name` is not present.

h-mayorquin replied:
Great. I did not know this trick with dictionaries. I did feel uncomfortable with the try-excepts block and I wanted to ask you if there a way around it. I will change.

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,131 @@
+        behavior_file_path = Path(session_path) / f"{session_id}.behavior.mat"
+
+        behavior_mat = read_matlab_file(str(behavior_file_path))["behavior"]
+        time_stamps = behavior_mat["timestamps"]
```
Doesn't look like you use these below after extracting them. Are they regular? (`from nwb_conversion_tools.utils.conversion_tools import check_regular_timestamps`, or something similarly named)

If they're regular you can forego all of it and use `starting_time` and `rate` as usual, if not we should use the timestamps instead.

h-mayorquin replied:
I was actually waiting to discuss this with you but it seems that all sampling is non-uniform.

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptal_utils.py:
```diff
@@ -0,0 +1,17 @@
+from scipy.io import loadmat as loadmat_scipy
+
+
+def read_matlab_file(file_path):
+    file_path = str(file_path)
+
+    try:
+        mat_file = loadmat_mat4py(str(file_path))
+    except:
+        try:
+            mat_file = loadmat_mat73(file_path)
+        except:
+            mat_file = loadmat_scipy(file_path)
+    return mat_file
```
Just as a heads up, while this kind of function will 'always work', each of those packages seems to have a different API structure for accessing internal values from the files, so any time you debug your behavior interfaces on a single file, you're debugging it for that API. But as we've sometimes seen, the version might change over a single session and so if you're not checking for that you might see missing keys simply because the file got loaded by a different reader.

h-mayorquin replied:
Yes. I think I will allocate some time to think better on how to consistently access matlab files in the same way. Right now I ran this for all the session and it seems to be working but I agree with you is a rather brittle solution.

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/convert_tingley_septal.py:
```diff
@@ -0,0 +1,101 @@
+
+# session_path_list = [Path("/shared/catalystneuro/Buzsaki/TingleyD/DT9/20170508_468um_36um_170508_102251")]
+
+counter = 0
+for session_path in session_path_list:
+    print("----------------")
+    print(session_path)
+    counter += 1
+    print(f"session {counter} of {len(session_path_list)}")
```
```suggestion
for counter, session_path in enumerate(session_path_list):
    print("----------------")
    print(session_path)
    print(f"session {counter} of {len(session_path_list)}")
```

h-mayorquin replied:
Yes, the reason I did not use an enumerate was because I was planning on removing it eventually.

CodyCBakerPhD commented:
Looking good, some comments above.

1) Any chance you could have some metadata added before our meeting? Last important piece of info before investigating files.

2) Did you get a change to check a session that has both Neuroscope and CellExplorer present at the same time? The current way the conversion is structured, you'll have a certain type of collision when they both attempt to write to the same `Units` table. But if they end up having the same spiking information, we'd want to default to using the CellExplorer (assuming it's not the unsupported version).

bendichter commented on buzsaki_lab_to_nwb/tingley_code_septal/convert_tingley_septal.py:
```diff
@@ -0,0 +1,106 @@
+            loadmat(session_info_matfile_path)
+            source_data.update(CellExplorerSorting=dict(spikes_matfile_path=str(spikes_matfile_path)))
+
+        except NotImplementedError:
+            warnings.warn("The CellExplorer data for this session is of a different version.")
```
Is this not handled by `read_matlab_file`?

CodyCBakerPhD replied:
Unfortunately, it'll take a bit more work than that to get that interface compatible with `hdf5storage` read patterns since it has slightly different indexing that we'd need to debug on a field-by-field basis. It's on the to-do list, though.

h-mayorquin commented:
Today:
* Added the trial intervals for the behavioral interface.
* Added the positional tracking error.
* Added the starting_time

In the last task there was something strange, one of the sessions is supposed to have been carried out on "20170229". However, as far as I know (and `datetime` in python is giving me that error) 2017 is not a leap year. I just decreased the day by one in that one.

h-mayorquin commented:
I tried to run the full conversion for some files and I am getting a new error that I did not see with the stubs.

```python
ValueError                                Traceback (most recent call last)
~/development/buzsaki-lab-to-nwb/buzsaki_lab_to_nwb/tingley_code_septal/convert_tingley_septal.py in <module>
    103         metadata=metadata,
    104         conversion_options=conversion_options,
--> 105         overwrite=True,
    106     )

~/development/nwb-conversion-tools/nwb_conversion_tools/nwbconverter.py in run_conversion(self, metadata, save_to_file, nwbfile_path, overwrite, nwbfile, conversion_options)
    177 
    178                 for interface_name, data_interface in self.data_interface_objects.items():
--> 179                     data_interface.run_conversion(nwbfile, metadata, **conversion_options.get(interface_name, dict()))
    180 
    181                 io.write(nwbfile)

~/development/nwb-conversion-tools/nwb_conversion_tools/datainterfaces/ecephys/baserecordingextractorinterface.py in run_conversion(self, nwbfile, metadata, stub_test, save_path, overwrite, **kwargs)
    139         validate(instance=conversion_opts, schema=conversion_opt_schema)
    140 
--> 141         self.writer_class.add_to_nwb(nwbfile, metadata, **conversion_opts)
    142         if save_path is not None:
    143             if overwrite:

~/development/nwb-conversion-tools/nwb_conversion_tools/utils/basesinwbephyswriter.py in add_to_nwb(self, nwbfile, metadata, **kwargs)
     68             return
     69         if self.recording is not None:
---> 70             self.add_recording()
     71             return
     72         if self.sorting is not None:

~/development/nwb-conversion-tools/nwb_conversion_tools/utils/si013nwbephyswriter.py in add_recording(self, segment_index)
    227 
    228     def add_recording(self, segment_index=0):
--> 229         super().add_recording(segment_index=segment_index)
    230         if self._conversion_ops["write_electrical_series"]:
    231             self.add_epochs()

~/development/nwb-conversion-tools/nwb_conversion_tools/utils/basesinwbephyswriter.py in add_recording(self, segment_index)
     46         self.add_devices()
     47         self.add_electrode_groups()
---> 48         self.add_electrodes()
     49         if self._conversion_ops["write_electrical_series"]:
     50             self.add_electrical_series(segment_index=segment_index)

~/development/nwb-conversion-tools/nwb_conversion_tools/utils/basenwbephyswriter.py in add_electrodes(self)
    289         for x in self.metadata["Ecephys"]["Electrodes"]:
    290             if x["name"] not in list(elec_columns):
--> 291                 raise ValueError(f'"{x["name"]}" not a property of se object, set it first and rerun')
    292             elec_columns[x["name"]]["description"] = x["description"]
    293 

ValueError: "shank_electrode_number" not a property of se object, set it first and rerun
```
I spent some time thinking about and I can try to understand it deeper but I wanted to check with you first. So far, I think the error comes from the following function:
https://github.com/catalystneuro/nwb-conversion-tools/blob/422740be88b01035a2732853b9cd18dd36463e38/nwb_conversion_tools/utils/basenwbephyswriter.py#L266

When `stub_test=True` this line returns:
`['group_name', 'group', 'shank_electrode_number', 'offset', 'gain']`
But when `stub_test=False` somehow the properties of `group_name` and `shank_electrode_number` are omitted. 
`['group', 'gain', 'offset']`

Importantly, the error persists even if I try to run the conversion with the data interface for the LFP so I don't think it is clash error.
The function itself is instantiated from the writer class `_get_channel_property_names` in the writer class `SI013NwbEphysWriter` which in turn uses the function `get_channel_property_names` from the `recorder` (NeuroscopeRecorder) in this case. 

I am not quite sure how changing the stub_test to True inhibits the process of the properties being read correctly. Have you seen something like this before?

CodyCBakerPhD commented:
@h-mayorquin Thanks for trying that; if you're getting ready for full runs, can you post some example stub files so we can inspect them prior to a complete full run?

> Have you seen something like this before?
I've seen many errors arise in going from `stub=False` to `stub=True`; but they're usually more complicated than this, which is a metadata problem somehow. Haven't seen this specific one before. Definitely something to come out of the recent refactor.

CodyCBakerPhD commented:
Here's why it doesn't show up in the stub: https://github.com/catalystneuro/nwb-conversion-tools/blob/422740be88b01035a2732853b9cd18dd36463e38/nwb_conversion_tools/utils/si013nwbephyswriter.py#L43-L48

The default behavior of stubbing (which previously had no effect on channels) changed to, indeed, subset the channels. I'll request that get reverted.

h-mayorquin commented:
> Here's why it doesn't show up in the stub: https://github.com/catalystneuro/nwb-conversion-tools/blob/422740be88b01035a2732853b9cd18dd36463e38/nwb_conversion_tools/utils/si013nwbephyswriter.py#L43-L48
> 
> The default behavior of stubbing (which previously had no effect on channels) changed to, indeed, subset the channels. I'll request that get reverted.

I don't think I understand fully. The error above appears when we go from the stub (stub=True) to the complete case (stub=False). That is, the properties are missing when we don't stub.

h-mayorquin commented:
All right,I just talked with Saksham today and he does not know -at least at first glance- what could be the source of the error. I will build a minimal example and raise and issue in conversion tools later today.

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalnwbconverter.py:
```diff
@@ -0,0 +1,134 @@
+            "acquired using an Intan RHD2000 system (Intan Technologies LLC) digitized with 20 kHz rate."
+        ),
+    ),
+    old_neuronexus_probe=dict(
+        name="Old neuronexus probe",
```
```suggestion
    old_neuronexus_probe=dict(
        name="Neuronexus probe (4 x 1)",
```

For consistency with the other names; you can re-use it in the case where you see multiple

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/convert_tingley_septal.py:
```diff
@@ -0,0 +1,114 @@
+
+    clu_matches_in_session = len(list(session_path.glob("*.clu*")))
+    res_matches_in_session = len(list(session_path.glob("*.res*")))
+    if clu_matches_in_session > 0 and res_matches_in_session > 0:
+        print("neuroscope sorted data available", True)
+        source_data.update(
+            NeuroscopeSorting=dict(
+                folder_path=str(session_path), keep_mua_units=False, xml_file_path=str(xml_file_path)
+            )
+        )
+        conversion_options.update(NeuroscopeSorting=dict(stub_test=stub_test))
+
+    if spikes_matfile_path.is_file():
+        try:
+            print("spikes and sessionInfo matlab files available", spikes_matfile_path.is_file())
+            loadmat(spikes_matfile_path)
+            loadmat(session_info_matfile_path)
+            # source_data.update(CellExplorerSorting=dict(spikes_matfile_path=str(spikes_matfile_path)))
+            source_data.update(CellExplorerSorting=dict(file_path=str(spikes_matfile_path)))
```
@h-mayorquin Can you nest this logic in the following way?

-> If CellExplorer data is present, add that to `source_data`
-> else, if Neuroscope data, add that to `source_data`

Under the broken `nwb-conversion-tools==0.10.3`, this was actually causing the `Units` columns to be corrupted with `Nan` values because the Neuroscope was overwriting in an append-like fashion.

Now that we're back to `nwb-conversion-tools==0.9.5`, it has a slightly safer behavior, but the Neuroscope is apparently overwriting the table completely, removing the custom column properties we want from CellExplorer.

h-mayorquin replied:
Makes sense, done.

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+            warnings.warn(f"Orientation data not found")
+
+        # States
+        module_name = "Neural states"
```
```suggestion
        module_name = "neural_states"
```
Reverting back to using underscores for all naming conventions now

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+
+            state_label_names = dict(WAKEstate="Awake", NREMstate="Non-REM", REMstate="REM", MAstate="MA")
+            sleep_state_dic = mat_file["SleepState"]["ints"]
+            table = TimeIntervals(name="Sleep states", description="Sleep state of the animal.")
```
```suggestion
            table = TimeIntervals(name="sleep_states", description="Sleep state of the animal.")
```

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+            compass_obj = CompassDirection(name=f"route centric")
+
+            spatial_series_object = SpatialSeries(
+                name=f"Orientation",
```
```suggestion
                name="orientation",
```

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+            ]
+            orientation_data = np.array(orientation_data)[..., 0]
+
+            module_name = "Orientation"
```
```suggestion
            module_name = "orientation"
```

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+            error_data = np.array([error for error in errorPerMarker])[..., 0]
+
+            spatial_series_object = SpatialSeries(
+                name=f"Error per marker",
```
```suggestion
                name="error_per_marker",
```

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+
+            spatial_series_object = SpatialSeries(
+                name=f"Error per marker",
+                description=f"Error per marker",
```
```suggestion
                description="Error per marker.",
```
I'd check for other unused f-strings as well.

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+        pos_obj = Position(name=f"{description}")
+
+        spatial_series_object = SpatialSeries(
+            name=f"Position",
```
```suggestion
            name=f"position",
```

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+            nwbfile.add_trial_column(name="trial_type", description="type of trial", data=trial_type_list)
+
+        # Position
+        module_name = "Position"
```
```suggestion
        module_name = "position"
```

bendichter replied:
can we try to stick with the pynwb modules for module names? e.g. for this can be put it in "behavior"? See best practices [here](https://www.nwb.org/best-practices/#naming-of-processing-modules)

CodyCBakerPhD replied:
That'd be fine; would `neural_states` count as `behavior` or `ecephys` (since they are inferred from the `ecephys` directly)

bendichter replied:
I would say ecephys

CodyCBakerPhD commented:
@h-mayorquin Just some minor renaming suggestions as per our verification of the standard practice in that regard. All the rest looks really good.

Only other thing I can mention is I do wish the orientation part didn't have to be done in a `try/except`; but I understand that maybe not all of the sub-fields (x,y,z,w) may exist, which makes it tricky to generalize that kind of code.

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+
+        spatial_series_object = SpatialSeries(
+            name=f"Position",
+            description=f"(x,y,z) coordinates tracking subject movement through",
```
```suggestion
            description=f"(x,y,z) coordinates tracking subject movement through.",
```

CodyCBakerPhD commented on buzsaki_lab_to_nwb/tingley_code_septal/tingleyseptalbehaviorinterface.py:
```diff
@@ -0,0 +1,167 @@
+        description = behavior_mat.get("description", "generic_position_tracking").replace("/", "-")
+        rotation_type = behavior_mat.get("rotationType", "non_specified")
+
+        pos_obj = Position(name=f"{description}")
```
```suggestion
        pos_obj = Position(name=f"{description.replace(" ", "_")}")
```
Noticed that some of the original names have spaces in them.

CodyCBakerPhD commented:
@h-mayorquin Looks great to me; going to go ahead and merge this just to make future requested changes easier to track.
